spec/truffle/interop/key_info_spec.rb

```diff
@@ -0,0 +1,201 @@
+      end
+    end
+    
+    it "returns :readable for all keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should include(:readable)
+      end
+    end
+    
+    it "returns :writable for all keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should include(:writable)
+      end
+    end
+    
+    it "does not return :writable for all keys if the hash is frozen" do
+      @hash.freeze
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should_not include(:writable)
+      end
+    end
+    
+    it "does not return :invocable" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should_not include(:invocable)
+      end
+    end
+    
+    it "does not return :internal for keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should_not include(:internal)
+      end
+    end
+    
+    it "returns nothing for a missing key" do
+      Truffle::Interop.key_info(@hash, :key_not_in_key_info_hash).should == []
+    end
+    
+    it "returns :existing for an instance variable that exists" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:existing)
+    end
+
+    it "does not return :existing for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should_not include(:existing)
+    end
+    
+    it "returns :readable for an instance variable" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:readable)
+    end
+    
+    it "does not return :readable for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should_not include(:readable)
+    end
+    
+    it "does not return :writable for an instance variable if the hash is frozen" do
+      @hash.freeze
+      Truffle::Interop.key_info(@hash, :@foo).should_not include(:writable)
+    end
+    
+    it "returns :writable for an instance variable" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:writable)
+    end
+    
+    it "returns :writable for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should include(:writable)
+    end
+    
+    it "returns :internal for an instance variable" do
+      Truffle::Interop.key_info(@hash, :@foo).should include(:internal)
+    end
+    
+    it "returns :internal for an instance variable that does not exist" do
+      Truffle::Interop.key_info(@hash, :@bar).should include(:internal)
+    end
+  
+  end
+  
+  describe "for a general object" do
+    
+    before :each do
+      @object = KeyInfoFixture.new
+    end
+
+    it "returns :existing for all readable or writable names" do
+      Truffle::Interop.key_info(@object, :ro).should include(:existing)
+      Truffle::Interop.key_info(@object, :rw).should include(:existing)
+      Truffle::Interop.key_info(@object, :wo).should include(:existing)
```

Comment:
I would think `:wo` is not existing (see the above case in the Hash and the ivar case below).
There is no method `wo` defined here (only `wo=`)

Reply:
I do think `wo` is `EXISTING` because someone has explicitly defined a setter for it. And the key is `wo`, not `wo=`, but we look at `wo=` to see if `wo` is defined as writable.

Reply:
The [JavaDoc](http://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/Message.html#KEY_INFO) mentions:

> KEY_INFO:
> This message also allows a fast check of existence of a property among Message.KEYS, the returned value is 0 iff the key does not exist. 

But `KEYS` returns the instance variables, not (accessor) methods.
I guess there is no single right answer here, and it's easy to change later if needed.

Reply:
The underlying problem is that Ruby's properties are encapsulated by methods that could do anything. It's hard to make it fit.